---
title: "SSU_qc_nonpool"
output: html_document
date: "2024-05-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
library(tibble)
```

```{r}
# read in taxonomy file
tax_f <- read.csv("../intermediates/ssu_fungal_taxonomy_table.csv", row.names=1)

abun_f <- read.csv("../intermediates/ssu_fungal_asv_table.csv", row.names=1)

```

Decontam

```{r}

library(decontam)

colnames(abun_f) <- gsub("pool", "Pool", colnames(abun_f))

snames <- data.frame(sample_names = colnames(abun_f)) 
snames$copy <- snames$sample_names
snames <- snames %>% separate(copy, c("Project", "Bromeliad", "SampType", "TechRep", "Locus", "other"), sep="\\.")

neg_samps <- snames$sample_names[grepl("NEG|PCR", snames$Bromeliad)]

# water filters have additional negatives, decontam separately from other samples
water_samps <- snames$sample_names[grepl("W", snames$SampType)]

all_watersamps <- c(neg_samps, water_samps)

water_negs <- water_samps[grepl("Blank", water_samps, ignore.case = TRUE)]

all_waternegs <- c(neg_samps, water_negs)

waterfilt_abun <- abun_f[,colnames(abun_f) %in% all_watersamps]
waterfilt_abun <- waterfilt_abun[rowSums(waterfilt_abun)>0,] # 2283 asvs, 196 samples
waterfilt_abun <- t(waterfilt_abun)

waterneg_bool <- rownames(waterfilt_abun) %in% all_waternegs


water_decontamed <- isContaminant(waterfilt_abun, method = "prevalence", neg = waterneg_bool)
table(water_decontamed$contaminant) # 2 ASVs identified as contaminants

w_decon_prev05 <- isContaminant(waterfilt_abun, method = "prevalence", neg = waterneg_bool, threshold = 0.5)
table(w_decon_prev05$contaminant) # 37 asvs

which_ones_water <- rownames(w_decon_prev05[w_decon_prev05$contaminant==TRUE,])

# 2246 asvs, 196 samples
decontamed_water_abun <- waterfilt_abun[,!(colnames(waterfilt_abun) %in% which_ones_water)]


## regular sample decontam
other_samp_abun <- abun_f[,!(colnames(abun_f) %in% water_samps)]
other_samp_abun <- other_samp_abun[rowSums(other_samp_abun)>0,] # 3241 asvs, 282 samples
other_samp_abun <- t(other_samp_abun)

reg_neg_bool <- rownames(other_samp_abun) %in% neg_samps


reg_decontamed <- isContaminant(other_samp_abun, method = "prevalence", neg = reg_neg_bool)
table(reg_decontamed$contaminant) # 8 ASVs identified as contaminants

reg_decon_prev05 <- isContaminant(other_samp_abun, method = "prevalence", neg = reg_neg_bool, threshold = 0.5)
table(reg_decon_prev05$contaminant) # 30 asvs

which_ones_reg <- rownames(reg_decon_prev05[reg_decon_prev05$contaminant==TRUE,])

# 3211 asvs, 282 samples
decontamed_other_samp_abun <- other_samp_abun[,!(colnames(other_samp_abun) %in% which_ones_reg)]


# take out negative samples? because they were decontamed differently
decontamed_other_samp_abun_noneg <- decontamed_other_samp_abun[!(rownames(decontamed_other_samp_abun) %in% neg_samps),]

decontamed_water_abun_noneg <- decontamed_water_abun[!(rownames(decontamed_water_abun) %in% neg_samps),]

# merge
all_decontamed_abun <- plyr::rbind.fill(as.data.frame(decontamed_other_samp_abun_noneg), 
                         as.data.frame(decontamed_water_abun_noneg))

all_decontamed_abun[is.na(all_decontamed_abun)] <- 0

rownames(all_decontamed_abun) <- c(rownames(decontamed_other_samp_abun_noneg), rownames(decontamed_water_abun_noneg))

all_decontamed_abun <- as.data.frame(t(all_decontamed_abun)) # 4734 asvs, 440 samples
all_decontamed_abun <- all_decontamed_abun[rowSums(all_decontamed_abun)>0,] # 439 samples, 4713 asvs
all_decontamed_abun <- all_decontamed_abun[,colSums(all_decontamed_abun)>0] 

# remove pcr positives
all_decontamed_abun_nopos <- all_decontamed_abun[,!(grepl("POS", colnames(all_decontamed_abun)))] # 438 samples

# remove water blanks
all_decontamed_abun_nopos <- all_decontamed_abun_nopos[,!(colnames(all_decontamed_abun_nopos) %in% water_negs)] # 423 samples
all_decontamed_abun_nopos <- all_decontamed_abun_nopos[rowSums(all_decontamed_abun_nopos)>0,] # 4693 asvs

```

Culling steps

```{r, echo=FALSE}
sampsums <- colSums(all_decontamed_abun_nopos)
asvsums <- rowSums(all_decontamed_abun_nopos)
```


First, she looks at the summary of sequencing depths:

```{r, echo=FALSE}
# summary of sequencing depths
summary(asvsums)

```


Then, checking how many asvs have no reads and what the minimum non-zero sequencing depth is:

```{r, echo=FALSE}

# check how many samples have no reads and what the minimum non-zero sequencing depth is
paste("Number of asvs with no reads:",sum(asvsums==0))
paste("Minimum non-zero sequencing depth:",sort(asvsums)[sum(asvsums==0)+1])

```

As well as whether the maximum may be an outlier:
```{r, echo=FALSE}
# is the maximum an outlier?
tail(sort(asvsums),5)
```


She said in her email that for rare asvs, she usually only excludes those seen in only one sample (unless building a network). The other common option would be something "very low like 10 reads." Here we can look at how many asvs we would exclude at either of those cutoffs.

Plot version:

```{r, echo=FALSE}

plot(sort(log1p(asvsums)), pch=19, xlab="Number of asvs", ylab="Log of Reads", main="asvs by read abundance with cutoffs set at 2, 5, and 10 reads")
abline(h=log1p(2), lty=2, col="blue") # this one?
abline(h=log1p(3), lty=2, col="green") 
abline(h=log1p(10), lty=2, col="magenta")
```


If we cull asvs with fewer than 10 reads:
```{r, echo=FALSE}
# hypothetical cutoff at 10 reads
paste("We would remove this many asvs:",sum(asvsums<10))
paste0("Which is ", round((sum(asvsums<10)/length(asvsums))*100,2), "% of asvs")
```

If we cull everything with fewer than 3 reads:
```{r, echo=FALSE}
# hypothetical cutoff at 6 reads
paste("We would remove this many asvs:", sum(asvsums<=2))
paste0("Which is ", round((sum(asvsums<=2)/length(asvsums))*100,2), "% of asvs")
```

### By Samples

Summary of sequencing depths by sample:

```{r, echo=FALSE}
# summary of sequencing depths
summary(sampsums)

```


How many samples have no reads and what the minimum non-zero sequencing depth is:

```{r, echo=FALSE}

# check how many samples have no reads and what the minimum non-zero sequencing depth is
paste("Number of Samples with no reads:",sum(sampsums==0))
paste("Minimum non-zero sequencing depth:",sort(sampsums)[sum(sampsums==0)+1])

```

Checking the maximum:
```{r, echo=FALSE}
# is the maximum an outlier?
tail(sort(sampsums),5)
```

Plot version:
Here we have cutoff lines at 1000 reads in pink, 100 reads in blue, and 10 reads in dark green.
```{r plots, echo=FALSE}
plot(sort(log1p(sampsums)), pch=19, xlab="Number of Samples", ylab="Log of Reads", main="Samples by read abundance with cutoffs set at 1000, 100, and 10 reads")
abline(h=log1p(150), lty=2, col="blue")
abline(h=log1p(1500), lty=2, col="magenta")
abline(h=log1p(300), lty=2, col="darkgreen") # this one
```


If we cull samples with 1000 or fewer reads:
```{r, echo=FALSE}
# hypothetical cutoff at 1000 reads
paste("We would remove this many Samples:",sum(sampsums<=300))
paste0("Which is ", round((sum(sampsums<300)/length(sampsums))*100,2), "% of Samples")
```

The 1400 read line appears to be around a "break" in the distribution.



```{r cull, echo=FALSE, message=FALSE}

# cull the abundance file (and then reduce the metadata and tax files by samples or asvs that are no longer there)

# original: 423 technical replicates, 4693 asvs

culled_abun <- all_decontamed_abun_nopos[,colSums(all_decontamed_abun_nopos)>300] # 5 samples removed, 418 remain
# remove asvs with 3 or fewer reads
culled_abun <- culled_abun[rowSums(culled_abun)>2,] # 541 ASVs removed, 4152 remain

saveRDS(culled_abun, "../intermediates/culled_fung_asv_table_decontamed_nonpooled.rds")


# cull tax table
culled_tax <- tax_f[tax_f$ASV %in% rownames(culled_abun),]

saveRDS(culled_tax, "../intermediates/culled_fung_tax_table_nonpooled.rds")


```


plots

```{r}
# final cutoffs
# remove samples with less than 300 reads
# remove asvs with less than or equal to than 3 reads

pdf("../figures/for_manuscript/bact_log_qc_cutoffs.pdf", width = 10, height=5)

par( mfrow= c(1,2) )

# by asv
plot(sort(log1p(asvsums)), pch=19, xlab="Number of ASVs", ylab="Log of Reads")
abline(h=log1p(3), lty=2, col="blue") # this one?

# by sample
plot(sort(log1p(sampsums)), pch=19, xlab="Number of Samples", ylab="Log of Reads")
abline(h=log1p(300), lty=2, col="magenta") # this one


dev.off()
```
















